std/machines/hash/keccakf16.asm

```diff
@@ -0,0 +1,608 @@
+    pol commit a_prime_prime_prime_0_0_limbs[4];
+
+    // Initially, the first step flag should be 1 while the others should be 0.
+    // builder.when_first_row().assert_one(local.step_flags[0]);
+    // for i in 1..NUM_ROUNDS {
+    //     builder.when_first_row().assert_zero(local.step_flags[i]);
+    // }
+    // for i in 0..NUM_ROUNDS {
+    //     let current_round_flag = local.step_flags[i];
+    //     let next_round_flag = next.step_flags[(i + 1) % NUM_ROUNDS];
+    //     builder
+    //         .when_transition()
+    //         .assert_eq(next_round_flag, current_round_flag);
+    // }
+
+    let step_flags: col[NUM_ROUNDS] = array::new(NUM_ROUNDS, |i| |row| if row % NUM_ROUNDS == i { 1 } else { 0 } );
+
+    // let main = builder.main();
+    // let (local, next) = (main.row_slice(0), main.row_slice(1));
+    // let local: &KeccakCols<AB::Var> = (*local).borrow();
+    // let next: &KeccakCols<AB::Var> = (*next).borrow();
+
+    // let first_step = local.step_flags[0];
+    // let final_step = local.step_flags[NUM_ROUNDS - 1];
+    // let not_final_step = AB::Expr::one() - final_step;
+
+    let first_step: expr = step_flags[0]; // aliasing instead of defining a new fixed column
+    let final_step: expr = step_flags[NUM_ROUNDS - 1];
+    // Equals to 1 for all non-last rows in all 24-row blocks. Forced to zero for the very last row of the submachine (if partial block).
+    // This is used to deactivate constraints that reference the next row, whenever we are at the latch row or the last row of the trace (so that we don't incorrectly cycle to the first row).
+    col fixed internal_row = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0]* + [0];
+
+    // // If this is the first step, the input A must match the preimage.
+    // for y in 0..5 {
+    //     for x in 0..5 {
+    //         for limb in 0..U64_LIMBS {
+    //             builder
+    //                 .when(first_step)
+    //                 .assert_eq(local.preimage[y][x][limb], local.a[y][x][limb]);
+    //         }
+    //     }
+    // }
+
+    array::zip(preimage, a, |p_i, a_i| first_step * (p_i - a_i) = 0);
+
+    // // The export flag must be 0 or 1.
+    // builder.assert_bool(local.export);
+
+    // force_bool(export);
+
+    // // If this is not the final step, the export flag must be off.
+    // builder
+    //     .when(not_final_step.clone())
+    //     .assert_zero(local.export);
+
+    // not_final_step * export = 0;
+
+    // // If this is not the final step, the local and next preimages must match.
+    // for y in 0..5 {
+    //     for x in 0..5 {
+    //         for limb in 0..U64_LIMBS {
+    //             builder
+    //                 .when(not_final_step.clone())
+    //                 .when_transition()
+    //                 .assert_eq(local.preimage[y][x][limb], next.preimage[y][x][limb]);
+    //         }
+    //     }
+    // }
+
+    array::map(preimage, |p| unchanged_until(p, internal_row));
```

Comment:
Another thing that should work is:
```rs
col fixed is_last = [0]* + [1];
// I think this can be inlined, as the new name basically describes how it is defined.
// let latch_row_or_last = final_step + is_last;
```

Because the block size is not a power of 2, both conditions will never be true at the same time (even if they were, it would work, because `unchanged_until` enforces the value to be unchanged as long as the condition is nonzero).

Plonky3 has a "native" `is_last` column, so with this approach we *could* save one fixed column (if we would detect that; currently, we'd create an `is_last` fixed column anyway).

Reply:
> I think `unchanged_until` means the opposite: If the flag is 0, it needs to be unchanged. So it should be `unchanged_until(p, 1 - internal_row)`, right?
> 
> But maybe then it makes sense to materialize the opposite:
> 
> ```rust
> col fixed latch_row_or_last = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]* + [1];
> ```

Yes you just catched a bug and the test passes after fixing it. :)

Reply:
> Another thing that should work is:
> 
> ```rust
> col fixed is_last = [0]* + [1];
> // I think this can be inlined, as the new name basically describes how it is defined.
> // let latch_row_or_last = final_step + is_last;
> ```
> 
> Because the block size is not a power of 2, both conditions will never be true at the same time (even if they were, it would work, because `unchanged_until` enforces the value to be unchanged as long as the condition is nonzero).
> 
> Plonky3 has a "native" `is_last` column, so with this approach we _could_ save one fixed column (if we would detect that; currently, we'd create an `is_last` fixed column anyway).



> Another thing that should work is:
> 
> ```rust
> col fixed is_last = [0]* + [1];
> // I think this can be inlined, as the new name basically describes how it is defined.
> // let latch_row_or_last = final_step + is_last;
> ```
> 
> Because the block size is not a power of 2, both conditions will never be true at the same time (even if they were, it would work, because `unchanged_until` enforces the value to be unchanged as long as the condition is nonzero).
> 
> Plonky3 has a "native" `is_last` column, so with this approach we _could_ save one fixed column (if we would detect that; currently, we'd create an `is_last` fixed column anyway).

This does looks cleaner to me and I'm doing this with the combined flag inlined. To clarify, we are not saving a column because we delete `internal_row` and add `is_last`.

Reply:
Yes, it's the same number of fixed columns! But, as mentioned, Plonky3 has a built-in `is_last` (because it is often needed to prevent wrapping), so we could remove this column again when compiling to the Plonky3 backend. But that's future work :)